### ISSUE TITLE
Can't play Noctis City when board is randomized

### ISSUE THREAD

**Issue by kberg** [terraforming-mars/terraforming-mars#4056]
Reported on Discord, when a player plays Noctis City with a randomized board, they are met with an error:

`Select a valid location: undefined is not land`

How peculiar! I think it has something to do with this line in `BoardBuilder.shuffle`:

```
[this.spaceTypes[land_id], this.spaceTypes[idx]] = [this.spaceTypes[idx], this.spaceTypes[land_id]];
```

Finding the true bug will probably be tricky, but for now I can patch.

**Comment by kberg** [comment]
This test reveals that empty space types happens fairly often, and only applies to reserved spaces:

```
  it('', () => {
    const spaces = new MultiSet<string>();
    for (let idx = 0; idx < 100000; idx++) {
      const seed = Math.random();
      board = OriginalBoard.newInstance({
        ...DEFAULT_GAME_OPTIONS,
        shuffleMapOption: true,
      },
      new Random(seed));
      for (const space of board.spaces) {
        if (space.spaceType === undefined) {
          console.log(`Bad seed ${seed}`);
          spaces.add(space.id);
        }
      }
    }
    if (spaces.size > 0) {
      console.log(spaces.toJSON());
    }
  });
```

```
Map(5) {constructor: ƒ, size: 5, 29 => 437, 14 => 401, 31 => 423, 21 => 451, 09 => 418}
```

That shows that the issue appears once in every 50 or so random boards.

(29, 14, 31, 21, 09 are all reserved spaces)

From `SpaceName`:
```
    ARSIA_MONS = '29',
    ASCRAEUS_MONS = '14',
    NOCTIS_CITY = '31',
    PAVONIS_MONS = '21',
    THARSIS_THOLUS = '09',
```

**Comment by kberg** [comment]
Fixed by #4061
